Use switch cases instead of for-loops over array-of-function pointers

d opened the pull request:
This patch is inspired by a change in #11285 but we've wholesale applied it to all of ORCA codebase. See each commit for more topical details.

## To reviewers
* The patch is organized into topical commits for convenience of review, but we plan to land the change as one squashed commit

* Most of the changes are fairly mundane, each "mundane" change roughly consists of 3 parts:
  1. Replace the look-up logic from a for-loop-over-array-of-function pointers to a straightforward switch case
  2. At every site of such a for-loop, there's always an array and a struct defined (with the lookup key and a function pointer). Remove the array, the "init" function, and the struct.
  3. This leaves behind a `typedef` that defines a shorthand of the function pointer. Remove the `typedef`.

* "Interesting" changes that require some nontrivial amount of braincells:
  1. "CXformSimplifySubquery": the logic is actually a lot more clear if you unnest the little array of 3 function pointers and express them in procedural code.
  2. "Remove unnecessary func ptrs in `CXformUtils::PexprBuildIndexPlan`": we have some uncalled-for complexity here passing around enums and function pointers, but the call sites always pass one constant value.

We validate that we leave no dead function pointer `typedef`s behind by sweeping the code base using the following query in `clang-query`:

```
match typedefNameDecl(hasType(anyOf(qualType(memberPointerType()),unless(qualType()))),unless(isExpansionInSystemHeader())).bind("typedef")
```

This is joint work with @chrishajas and @hardikar , the "interesting" logic is courtesy of @hardikar .

zellerh commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -359,23 +264,220 @@ CTranslatorDXLToPlStmt::TranslateDXLOperatorToPlan(
+		}
+		case EdxlopPhysicalExternalScan:
+		{
+			plan = TranslateDXLTblScan(dxlnode, output_context,
```
Some more opportunities to combine cases in this statement.

d replied:
Thanks! Will do!

zellerh commented on src/backend/gporca/libnaucrates/src/operators/CDXLOperatorFactory.cpp:
```diff
@@ -2435,88 +2435,152 @@ CDXLOperatorFactory::GetDatumVal(CDXLMemoryManager *dxl_memory_manager,
-		 &CDXLOperatorFactory::GetDatumStatsLintMappable},
+		case GPDB_CHAR:
+		{
+			return GetDatumStatsLintMappable(dxl_memory_manager, attrs,
```
More opportunities for merging cases.

zellerh reviewed (approved):
Thanks to all of you for this big improvement!!

d commented on src/backend/gpopt/translate/CTranslatorDXLToScalar.cpp:
```diff
@@ -1541,47 +1577,49 @@ Expr *
+		}
+		case CDXLDatum::EdxldatumStatsDoubleMappable:
+		{
+			return (Expr *) TranslateDXLDatumGenericToScalar(datum_dxl);
```
Cool will do!

d commented on src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:
```diff
@@ -3003,37 +3003,40 @@ CTranslatorQueryToDXL::TranslateFromClauseToDXL(Node *node)
+			}
+			case RTE_RELATION:
+			{
+				return CTranslatorQueryToDXL::TranslateRTEToDXLLogicalGet(
```
Oh yes this is my oversight, fixing it right now

d commented:
@zellerh I've pushed a change that addresses your feedback on
1. Removing namespace qualification before member function calls
2. Use more fall-through `case`s

In the 4 files you commented on. I've left the other feedback unresolved for now.

hardikar commented on src/backend/gporca/libgpopt/src/xforms/CXformSimplifySubquery.cpp:
```diff
@@ -229,49 +218,69 @@ CXformSimplifySubquery::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
-						pexprInput->Pop()->Eopid());
+	// the last entry is used to replace existential subqueries with count(*)
+	// after quantified subqueries have been replaced in the input expression
+	pexprResult = FSimplifySubquery(mp, pexpr, FSimplifyExistential,
```
Yes we should. Looks like that was missed when we rebased the branch before opening the PR. What's surprising is that no test failed.

d replied:
Yeah I *think* I fat-fingered that conflict resolution during rebase :-(

hardikar replied:
No, @d, it happened on my computer.

zellerh commented on src/backend/gporca/libgpopt/src/xforms/CXformSimplifySubquery.cpp:
```diff
@@ -236,7 +236,7 @@ CXformSimplifySubquery::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
 	// the last entry is used to replace existential subqueries with count(*)
 	// after quantified subqueries have been replaced in the input expression
-	pexprResult = FSimplifySubquery(mp, pexpr, FSimplifyExistential,
+	pexprResult = FSimplifySubquery(mp, pexprResult, FSimplifyExistential,
```
Doesn't this need a nullptr check? Or just move this code inside the if statement on line 232?

hardikar replied:
Yes it does! Fix pushed. Thanks for pointing that out.

dgkimura commented on src/include/gpopt/translate/CTranslatorQueryToDXL.h:
```diff
@@ -269,7 +269,7 @@ class CTranslatorQueryToDXL
 
 	// throws an exception when RTE kind not yet supported
-	void UnsupportedRTEKind(RTEKind rtekind) const;
+	[[noreturn]] void UnsupportedRTEKind(RTEKind rtekind) const;
```
Oo, nice change!

dgkimura reviewed (approved):
Nice change!  I like it as it makes debugging much easier by requiring less conditional breakpoints.

Possibly one downside is that by changing to switch/case I _think_ the compiler may have to add extra jump statements for each case in the switch?  Whereas with array it can just load the offset into a register and call.  But either way, that seems like a small price to pay for friendlier debugging.

d commented:
> Whereas with array it can just load the offset into a register and call.

This is only true if we use the array as follows:

```c++
void DoFoo(EopId opid) {
  return m_fp[opid](arg1, arg2);
}
```

The majority of our uses are much more like this:

```c++
void DoBar(EopId opid) {
  for (auto mapper: m_struct_containing_fp) {
    if (mapper.opid == opid)
      return (*mapper.fp)(arg1, arg2);
  }
}
```

This pattern is bummer for many reasons:
1. Worse readability: the called functions' call hierarchy will not lead to `DoBar`
2. Worse generated code: the lookup is gonna be no better than a switch, and the call is a single indirect branch (meaning it's gonna compile to some variant of "load and then branch"). This typically kills the branch predictor because different `EopId` values correspond to the same instruction pointer.
3. In contrast, each call in a switch case is a direct branch, meaning that it's friendlier to the branch predictor, _and_ it's easier to speculate for the processor

dgkimura commented:
> > Whereas with array it can just load the offset into a register and call.
> 
> This is only true if we use the array as follows:
> 
> ```c++
> void DoFoo(EopId opid) {
>   return m_fp[opid](arg1, arg2);
> }
> ```
> 
> The majority of our uses are much more like this:
> 
> ```c++
> void DoBar(EopId opid) {
>   for (auto mapper: m_struct_containing_fp) {
>     if (mapper.opid == opid)
>       return (*mapper.fp)(arg1, arg2);
>   }
> }
> ```

Ah, yeah.  You're totally right.  Seems like with some extra finagling the for loop may have been removable.  But since it wasn't,  I can't see any reason for why function pointers where used..

d commented:
One understated benefit of moving away from array-of-function-pointers, apart from how the for-loop kills performance, is opportunities at simplifying and improving our code: the array-of-function-pointers approach imposes a rigid assumption about each function that needs to be called. This often led to:

1. unused arguments: if you need to add a parameter to one of the functions, signatures change for _all_ functions;
2. unnecessary _non const_ functions: again, you cannot make just one function `const` if the rest of them aren't;
3. and unnecessary `this` pointer: ditto, you can't just make one function `static` if it needs to fit into the function pointer type
4. overall, it inhibits automated tools that help simplify the code like clang-tidy checks `readability-convert-member-functions-to-static` (https://clang.llvm.org/extra/clang-tidy/checks/readability-convert-member-functions-to-static.html) and `readability-make-member-function-const` (https://clang.llvm.org/extra/clang-tidy/checks/readability-make-member-function-const.html)

d commented:
@hardikar the tip fixup commit LGTM. I think we can start drafting a message for the final squashed commit, WDYT?

d commented:
Proposed commit message:

```
Use switch cases, not loops over function pointers.

Context: ORCA has a pattern of pattern matching function calls like the
following:

Expr *TranslateDXLToScalar(Edxlopid eopid, const CDXLNode *dxlnode) {
  for (STranslatorElem elem : m_translators) {
    if (eopid == elem.eopid) {
      translate_func = elem.translate_func;
      return (this->*translate_func)(dxlnode);
    }
  }
}

This lookup-by-for-loop pattern is a bummer for many reasons:

1. Worse readability: the called functions' call hierarchy will not lead
to the lookup function

2. Worse generated code: the lookup is gonna be no better than a switch,
and the call is a single indirect branch (meaning it's gonna compile to
some variant of "load and then branch").

In this patch, we replace of of the function pointer lookup code with
simple, straightforward switch cases that perform direct function calls.


Most of the changes are fairly mundane, each "mundane" change roughly consists of 3 parts:

1. Replace the look-up logic from a for-loop-over-array-of-function
pointers to a straightforward switch case

2. At every site of such a for-loop, there's always an array and a
struct defined (with the lookup key and a function pointer). Remove the
array, the "init" function, and the struct.

3. This leaves behind a typedef that defines a shorthand of the function
pointer. Remove the typedef.

"Interesting" changes that require some nontrivial amount of braincells:

1. "CXformSimplifySubquery": the logic is actually a lot more clear if
you unnest the little array of 3 function pointers and express them in
procedural code.

2. "Remove unnecessary func ptrs in CXformUtils::PexprBuildIndexPlan":
we have some uncalled-for complexity here passing around enums and
function pointers, but the call sites always pass one constant value.

One understated benefit of moving away from array-of-function-pointers,
apart from how the for-loop kills performance, is opportunities at
simplifying and improving our code: the array-of-function-pointers
approach imposes a rigid assumption about each function that needs to be
called. This often led to:

1. unused arguments: if you need to add a parameter to one of the
functions, signatures change for all functions;

2. unnecessary non-const functions: again, you cannot make just one
function const if the rest of them aren't; and unnecessary this pointer:
ditto, you can't just make one function static if it needs to fit into
the function pointer type

3. overall, it inhibits automated tools that help simplify the code like
clang-tidy checks readability-convert-member-functions-to-static
(https://clang.llvm.org/extra/clang-tidy/checks/readability-convert-member-functions-to-static.html)
and readability-make-member-function-const
(https://clang.llvm.org/extra/clang-tidy/checks/readability-make-member-function-const.html)

This patch is inspired by a change in greenplum-db/gpdb#11285 but we've
wholesale applied it to all of ORCA codebase.

```
